The asm here results in incorrect code generation when passing
single precision floats with Clang.

Therefore, implement the storing without assembly. The D extension
specification states:

When multiple floating-point precisions are supported, then valid values of narrower n-bit types,
n < FLEN, are represented in the lower n bits of an FLEN-bit NaN value, in a process termed
NaN-boxing. The upper bits of a valid NaN-boxed value must be all 1s.

That means we can just set the value to all ones, copy over the float
and finally store it in an actual float type (via memcpy to avoid
violating aliasing rules).

In case of return values, we have the data loaded from the register
into memory and we know we need just the 32 bits so simply take them.

diff -ruN a/src/riscv/ffi.c b/src/riscv/ffi.c
--- a/src/riscv/ffi.c	2026-08-09 00:12:21.000000000 +0200
+++ b/src/riscv/ffi.c	2026-10-01 08:33:11.993447288 +0200
@@ -183,12 +183,14 @@
             return;
 #elif ABI_FLEN >= 32
         case FFI_TYPE_FLOAT:
-            __asm__("" : "=f"(cb->aregs->fa[cb->used_float++].f) : "0"(*(float *)data));
+            value = ~(size_t)0;
+            memcpy(&value, data, sizeof(float));
+            memcpy(&cb->aregs->fa[cb->used_float++], &value, sizeof(ABI_FLOAT));
             return;
 #endif
 #if ABI_FLEN >= 64
         case FFI_TYPE_DOUBLE:
-            __asm__("" : "=f"(cb->aregs->fa[cb->used_float++].d) : "0"(*(double *)data));
+            memcpy(&cb->aregs->fa[cb->used_float++], data, sizeof(double));
             return;
 #endif
         default: FFI_ASSERT(0); break;
@@ -210,12 +212,12 @@
             return;
 #elif ABI_FLEN >= 32
         case FFI_TYPE_FLOAT:
-            __asm__("" : "=f"(*(float *)data) : "0"(cb->aregs->fa[cb->used_float++].f));
+            memcpy(data, &cb->aregs->fa[cb->used_float++], sizeof(float));
             return;
 #endif
 #if ABI_FLEN >= 64
         case FFI_TYPE_DOUBLE:
-            __asm__("" : "=f"(*(double *)data) : "0"(cb->aregs->fa[cb->used_float++].d));
+            memcpy(data, &cb->aregs->fa[cb->used_float++], sizeof(double));
             return;
 #endif
     }
